### ISSUE TITLE
Migration does not detect custom SQLDeclaration change

### ISSUE THREAD

**Issue by michnovka** [doctrine/dbal#5308]
### Bug Report
If declaring custom `Type` and its `getSQLDeclaration()` method changes, no migration is generated.

|    Q        |   A
|------------ | ------
| Version     | 3.3.2

#### Summary

Implement ENUM like in the official [docs](https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/cookbook/mysql-enums.html) Then if you add/remove new ENUM element, the return value of `getSQLDeclaration()` changes. But this is not detected by Comparer and no migration is generated

#### How to reproduce

Create type:
```php
<?php
namespace MyProject\DBAL;

use Doctrine\DBAL\Types\Type;
use Doctrine\DBAL\Platforms\AbstractPlatform;

class EnumVisibilityType extends Type
{
    const ENUM_VISIBILITY = 'enumvisibility';
    const STATUS_VISIBLE = 'visible';
    const STATUS_INVISIBLE = 'invisible';

    public function getSQLDeclaration(array $fieldDeclaration, AbstractPlatform $platform)
    {
        return "ENUM('visible', 'invisible')";
    }

    public function convertToPHPValue($value, AbstractPlatform $platform)
    {
        return $value;
    }

    public function convertToDatabaseValue($value, AbstractPlatform $platform)
    {
        if (!in_array($value, array(self::STATUS_VISIBLE, self::STATUS_INVISIBLE))) {
            throw new \InvalidArgumentException("Invalid status");
        }
        return $value;
    }

    public function getName()
    {
        return self::ENUM_VISIBILITY;
    }

    public function requiresSQLCommentHint(AbstractPlatform $platform)
    {
        return true;
    }
}
```

2. Create Entity with column of this type

```php
<?php
/** @Entity */
class Article
{
    /** @Column(type="enumvisibility") */
    private $status;
}
```

3. Migrate and save DB schema change.
4. Change `getSQLDeclaration` to
```php
        return "ENUM('visible', 'invisible','halfvisible')";
```
5. Do migrations:diff

#### Expected behaviour

New migration is created reflecting the change

**Comment by michnovka** [comment]
The issue is that `AbstractPlatform::colmnsEqual` returns updated `$this->getColumnDeclarationSQL('', $column1Array)` also for `$this->getColumnDeclarationSQL('', $column2Array)`. Any idea where we can check for the actual current column definition in DB?

**Comment by michnovka** [comment]
I dug real deep into the code and I came to conclusion that this is unfixable until we implement ENUM as valid DB type in Doctrine. The issue is that in the end `DBAL\Schema\MySQLSchemeManager::_getPortableTableColumnDefinition` [checks the column type](https://github.com/doctrine/dbal/blob/baf6efed70e46f4d979e5dd42f804007e75f12fb/src/Schema/MySQLSchemaManager.php#L145) and determines type-specific properties (length for text, precision for doubles etc.) This all is then passed to a `Column` object in `$options` array.
The proper way would be to add `$options['enum_options']` and parse ENUM options into it. Then we can have column diff to actually check for change in these and properly report ENUM type as changed.

**Comment by morozov** [comment]
I don't believe schema migrations could ever work correctly when the column definition is specified manually. See https://github.com/doctrine/dbal/issues/3647#issuecomment-519701787. May be related to https://github.com/doctrine/dbal/issues/5223.

**Comment by morozov** [comment]
After digging a bit more through the codebase, I believe the types defined as the [documentation](https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/cookbook/mysql-enums.html#solution-2-defining-a-type) suggests will never work properly with the current design.

On the one hand, a type can declare an arbitrary DDL to represent itself. E.g.: https://github.com/doctrine/dbal/blob/c826c18da3b5631d6a57d5c93ae2248f0e8cfbbe/tests/Functional/Schema/MySQL/PointType.php#L15-L18

Besides that, a type can use arbitrary undocumented properties that influence the resulting DDL. E.g. (https://github.com/doctrine/dbal/issues/5307):

https://github.com/doctrine/dbal/blob/9b155f10294e2e1700bc15f9112de4862a8fb1e8/src/Platforms/AbstractMySQLPlatform.php#L154-L161

This way, the API for building type DDL is quite flexible, composable, and appealing to use.

On the other hand, the logic of type introspection is absolutely rigid and hard-coded:
https://github.com/doctrine/dbal/blob/9b155f10294e2e1700bc15f9112de4862a8fb1e8/src/Schema/MySQLSchemaManager.php#L113-L226

Not only that it's non-extensible, but it's also absolutely incoherent with its DDL counterpart which makes it prone to the issues like this.
